pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java

```diff
@@ -231,185 +184,149 @@ public void initialize(ServiceConfiguration conf, PulsarResources pulsarResource
+                    if (canConsumer) {
+                        return CompletableFuture.completedFuture(true);
                     }
-                    allowTheSpecifiedActionOpsAsync(namespaceName, role, authenticationData, AuthAction.produce)
-                            .whenComplete((produceAuthorized, ex) -> {
-                                if (ex == null) {
-                                    finalResult.complete(produceAuthorized);
-                                } else {
-                                    if (log.isDebugEnabled()) {
-                                        log.debug("Namespace [{}] Role [{}] exception occurred while trying to check "
-                                                + "Produce permission. {}", namespaceName, role, ex.getCause());
-                                    }
-                                    finalResult.completeExceptionally(ex.getCause());
-                                }
-                            });
+                    return allowTheSpecifiedActionOpsAsync(namespaceName, role, authenticationData, AuthAction.produce);
                 });
-
-        return finalResult;
     }
 
     private CompletableFuture<Boolean> allowTheSpecifiedActionOpsAsync(NamespaceName namespaceName, String role,
                                                                        AuthenticationDataSource authenticationData,
                                                                        AuthAction authAction) {
-        CompletableFuture<Boolean> permissionFuture = new CompletableFuture<>();
-        try {
-            pulsarResources.getNamespaceResources().getPoliciesAsync(namespaceName).thenAccept(policies -> {
-                if (!policies.isPresent()) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("Policies node couldn't be found for namespace : {}", namespaceName);
-                    }
-                } else {
-                    Map<String, Set<AuthAction>> namespaceRoles = policies.get()
-                            .auth_policies.getNamespaceAuthentication();
-                    Set<AuthAction> namespaceActions = namespaceRoles.get(role);
-                    if (namespaceActions != null && namespaceActions.contains(authAction)) {
-                        // The role has namespace level permission
-                        permissionFuture.complete(true);
-                        return;
-                    }
+        return pulsarResources.getNamespaceResources().getPoliciesAsync(namespaceName).thenApply(policies -> {
+            if (!policies.isPresent()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Policies node couldn't be found for namespace : {}", namespaceName);
+                }
+            } else {
+                Map<String, Set<AuthAction>> namespaceRoles = policies.get()
+                        .auth_policies.getNamespaceAuthentication();
```

Comment:
I believe that we could think more about this objects (like auth_policies  here) that can be null

Reply:
Hi @eolivelli . Seems safe, because it is final. But thanks for your reminder, I will check the other fields.